src/htm/encoders/SimHashDocumentEncoder.cpp

```diff
@@ -56,63 +56,153 @@ namespace htm {
+      NTA_CHECK(parameters.charFrequencyCeiling == 0)
+        << "Argument 'charFrequencyCeiling' requires argument 'tokenSimilarity'.";
+      NTA_CHECK(parameters.charFrequencyFloor == 0)
+        << "Argument 'charFrequencyFloor' requires argument 'tokenSimilarity'.";
+      if ((parameters.charFrequencyCeiling > 0) && (parameters.charFrequencyFloor > 0)) {
+        NTA_CHECK(parameters.charFrequencyCeiling > parameters.charFrequencyFloor)
+          << "Argument 'charFrequencyCeiling' should be greater than argument 'charFrequencyFloor'.";
+      }
+    }
+    if (!parameters.vocabulary.size()) {
+      NTA_CHECK(!parameters.encodeOrphans)
+        << "Argument 'encodeOrphans' requires argument 'vocabulary'.";
+    }
 
     // Make local copy of arguments for processing
     args_ = parameters;
 
     // process: handle 'sparsity' param instead of 'activeBits' param
-    if(args_.sparsity > 0.0f) {
+    if (args_.sparsity > 0.0f) {
       NTA_CHECK((args_.sparsity >= 0.0f) && (args_.sparsity <= 1.0f))
         << "Argument 'sparsity' must be a float in the range 0.0-1.0.";
       NTA_CHECK(args_.size > 0u)
         << "Argument 'sparsity' requires that the 'size' also be given.";
       args_.activeBits = (UInt) round(args_.size * args_.sparsity);
     }
 
+    // Process: handle internal case insensitivity needs
+    if (!args_.caseSensitivity) {
+      // exclusions => case insensitive
+      if (!args_.excludes.empty()) {
+        std::vector<std::string> excludesLower;
+        for (auto& token : args_.excludes) {
+          transform(token.begin(), token.end(), token.begin(), ::tolower);
+          excludesLower.push_back(token);
+        }
+        args_.excludes = excludesLower;
+      }
+
+      // vocabulary => case insensitive
+      if (args_.vocabulary.size()) {
+        std::map<std::string, UInt> vocabLower;
+        for (const auto& pair : args_.vocabulary) {
+          std::string token = pair.first;
+          const UInt weight = pair.second;
+          transform(token.begin(), token.end(), token.begin(), ::tolower);
+          vocabLower[token] = weight;
+        }
+        args_.vocabulary = vocabLower;
+      }
+    }
+
     // Post-processing, Sanity check the parameters
     NTA_CHECK(args_.size > 0u);
     NTA_CHECK(args_.activeBits > 0u);
     NTA_CHECK(args_.activeBits < args_.size);
 
     // Initialize parent class with finalized params
-    BaseEncoder<std::map<std::string, UInt>>::initialize({ args_.size });
+    BaseEncoder<std::vector<std::string>>::initialize({ args_.size });
   } // end method initialize
 
   /**
    * Encode (Main calling style)
    * @see SimHashDocumentEncoder.hpp
-   * @see encode(const std::vector<std::string> input, SDR &output)
    * @see encode(std::string input, SDR &output)
    */
-  void SimHashDocumentEncoder::encode(const std::map<std::string, UInt> input, SDR &output)
+  void SimHashDocumentEncoder::encode(const std::vector<std::string> input, SDR &output)
```

Comment:
nit, just a question. Do we need to provide the `vector<string>` option, or could we limit ourselves to just the `encode(string)` version?

Reply:
I'm pretty sure I'll use this myself, but it's also no biggie, lemme know. :)

Reply:
fine to keep, if you have a use for that :+1: